### ISSUE TITLE
Incorrect order when outputting errors inside generated def

### ISSUE THREAD

**Issue by paulcsmith** [crystal-lang/crystal#4699]
I searched and didn't see anything about this so I think this is not a duplicate.

Let's say I have something like this

```crystal
macro gen_tag_helpers
  {% for tag in [:a, :img, :hr, :div] %}
    def {{ tag.id }}(text)
      puts text.upcase
    end
  {% end %}
end

gen_tag_helpers

not_string = 1
# This will blow up because `1` does not implement `upcase`
img(not_string)
```

The output is this:

```
Error in macro_errors.cr:12: instantiating 'img(Int32)'

img(not_string)
^~~

in macro_errors.cr:9: expanding macro

gen_tag_helpers
^

in macro 'gen_tag_helpers' /Users/paulcsmith/learning/crystal-playground/macro_errors.cr:1, line 7:

   1.
   2.     def a(text)
   3.       puts text.upcase
   4.     end
   5.
   6.     def img(text)
>  7.       puts text.upcase
   8.     end
   9.
  10.     def hr(text)
  11.       puts text.upcase
  12.     end
  13.
  14.     def div(text)
  15.       puts text.upcase
  16.     end
  17.

undefined method 'upcase' for Int32
Rerun with --error-trace to show a complete error trace.
```

The problem here is that I would expect this to be last 

```
img(not_string)
^~~
```

This may be how the compiler sees it, but as a developer I expect to see macro output go above the macro debug output. This is especially problematic when the macro generates a lot of code and you have to scroll up to see what code caused the failure.

**Comment by straight-shoota** [comment]
The most important stuff is usually put first, not last. Location in code, error messages, code excerpt... and only after that comes the callstack or as in this case contextual information like expanded macro code.
If you're running the compiler in a terminal, that would mean to burden yourself with scrolling upwards to read the top of the error output if there is a lot. But this is merely a question of the user interface. In a terminal you could pipe the output through a pager to read from the top.

**Comment by paulcsmith** [comment]
Usually that is the case, but it's not the case for compiler errors in crystal. They're opposite of a normal stack trace. I'm also fine with it being first like most stacktraces, but it should be consistent. 

Here is an example:

```crystal
class User
  def normalize_name(name)
    name.upcase
  end
end

name = User.new.normalize_name(1)
```

And the errors, notice that the order is opposite the normal order. I actually prefer this and think it should be consistent with macros as well.

```
Error in compile_errors.cr:7: instantiating 'User#normalize_name(Int32)'

name = User.new.normalize_name(1)
                ^~~~~~~~~~~~~~

in compile_errors.cr:3: undefined method 'upcase' for Int32

    name.upcase
         ^~~~~~

Rerun with --error-trace to show a complete error trace.
```

**Comment by asterite** [comment]
Duplicate of #4700

The order is **exactly the same**, be it a method or a macro.

**Comment by paulcsmith** [comment]
@asterite Sorry I was unclear in my descriptions and titles. This issue is not a duplicate of #4700, though they both stem from the same root problem of unclear (IMO) compiler errors when the error is in generate code from a macro.

1. The suggestion in #4700 was to only show a few lines surrounding the error when there is a lot of generated code, rather than *everything*. Macros that generate a lot of code make it really hard to find where in *my app code* the problem lies, because I have to slowly scroll up until I get to the app code that I wrote.
1. I incorrectly labelled this issue (#4699) as "incorrect order" when the root problem is that it's unexpected/less helpful order. I see your point now that technically the macro is the last thing to be called so it comes at the bottom so it's not a bug. Though that is technically correct, it makes debugging very difficult. I have to go through a bunch of macro output to see what code *I* wrote that is the problem. Usually the last line of compiler output is what *I* wrote, but when calling a macro in another library/stdlib it's code that I didn't write. So I think that the code I called should be last. Otherwise I have to slowly scroll through the stack trace until I found the code that I wrote that was the problem. 

Does that make sense? So neither of these two things are bug, but implementing one or both would help improve finding what code I wrote that caused the issue.

So maybe the best bet is to always put the user's app code (not the generate code) at the bottom of the output? Maybe separate and with a different color to make it clear what code I wrote caused the issue?

Maybe this?

```
Error in macro_errors.cr:12: instantiating 'img(Int32)'

img(not_string)
^~~

in macro_errors.cr:9: expanding macro

gen_tag_helpers
^

in macro 'gen_tag_helpers' /Users/paulcsmith/learning/crystal-playground/macro_errors.cr:1, line 7:

   1.
   2.     def a(text)
   3.       puts text.upcase
   4.     end
   5.
   6.     def img(text)
>  7.       puts text.upcase
   8.     end
   9.
  10.     def hr(text)
  11.       puts text.upcase
  12.     end
  13.
  14.     def div(text)
  15.       puts text.upcase
  16.     end
  17.

undefined method 'upcase' for Int32

---------------------------------------------------
# And maybe highlight the code in red?
undefined method 'upcase' for Int32 in file xxx/xxx.cr

12. img(not_string)
^~~

Rerun with --error-trace to show a complete error trace.
```
